waspc/src/Wasp/Analyzer/TypeChecker/TypeError.hs

```diff
@@ -82,22 +82,35 @@ data TypeCoercionErrorReason e
     ReasonDictWrongKeyType String e
   deriving (Eq, Show)
 
-getTypeCoercionErrorMessageAndCtx :: (Type -> TypedExpr -> String) -> TypeCoercionError -> (String, Ctx)
-getTypeCoercionErrorMessageAndCtx getUncoercableTypesMsg (TypeCoercionError (WithCtx ctx texpr) t reason) =
+second :: (b -> d) -> (a, b, c) -> (a, d, c)
+second f (x, y, z) = (x, f y, z)
+
+extractTypeErrorMessagesAndCtx :: (Type -> TypedExpr -> String) -> TypeCoercionError -> (String, [String], Ctx)
+extractTypeErrorMessagesAndCtx getUncoercableTypesMsg (TypeCoercionError (WithCtx ctx texpr) t reason) =
   case reason of
     ReasonList e ->
-      first (("For list element:\n" ++) . indent 2) $
-        getTypeCoercionErrorMessageAndCtx getUncoercableTypesMsg e
+      second ("-> In list element" :) $ extractTypeErrorMessagesAndCtx getUncoercableTypesMsg e
     ReasonDictWrongKeyType key e ->
-      first ((("For dictionary field '" ++ key ++ "':\n") ++) . indent 2) $
-        getTypeCoercionErrorMessageAndCtx getUncoercableTypesMsg e
-    ReasonDictNoKey key -> ("Missing required dictionary field '" ++ key ++ "'", ctx)
-    ReasonDictExtraKey key -> ("Unexpected dictionary field '" ++ key ++ "'", ctx)
+      second (("-> In dictionary field '" ++ key ++ "'") :) $ extractTypeErrorMessagesAndCtx getUncoercableTypesMsg e
+    ReasonDictNoKey key -> ("-> Missing required dictionary field '" ++ key ++ "'", [], ctx)
+    ReasonDictExtraKey key -> ("-> Unexpected dictionary field '" ++ key ++ "'", [], ctx)
     ReasonDecl -> uncoercableTypesMsgAndCtx
     ReasonEnum -> uncoercableTypesMsgAndCtx
     ReasonUncoercable -> uncoercableTypesMsgAndCtx
   where
-    uncoercableTypesMsgAndCtx = (getUncoercableTypesMsg t texpr, ctx)
+    uncoercableTypesMsgAndCtx = (getUncoercableTypesMsg t texpr, [], ctx)
```

Comment:
Since we give the "real" message special treatment (I.e., we display it first, in an order inconsistent with the rest of the messages), I thought that it probably shouldn't be bundled in the same array as the rest of them.

Also, I decided to put the additional messages in an array (as opposed to a string) and handle the indentation separately. I felt weird initializing the second member of the tuple with `""`. It also caused the `indent 2` part to add trailing spaces.

With that said, there is probably a better way to do most of this. Looking forward to the comments :)